Move HashCountedSet to WeakHashCountedSet in StyleGeneratedImage

chirags27 opened the pull request:
#### 4bfd15d8cdc7d90dc7674237addf934a3d5c7be4
<pre>
Move HashCountedSet to WeakHashCountedSet in StyleGeneratedImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=254835">https://bugs.webkit.org/show_bug.cgi?id=254835</a>
rdar://107480319

Reviewed by Chris Dumez.

Generated images should use a Weak container to keep track of
RenderElements so that we don&apos;t trigger UAF issues.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore:: const):
* Source/WebCore/rendering/style/StyleCanvasImage.cpp:
(WebCore::StyleCanvasImage::image const):
(WebCore::StyleCanvasImage::canvasChanged):
(WebCore::StyleCanvasImage::canvasResized):
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
(WebCore::StyleCrossfadeImage::imageChanged):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::imageChanged):
* Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
(WebCore::StyleGeneratedImage::addClient):
(WebCore::StyleGeneratedImage::removeClient):
(WebCore::StyleGeneratedImage::hasClient const):
* Source/WebCore/rendering/style/StyleGeneratedImage.h:
(WebCore::StyleGeneratedImage::clients const):
(WebCore::StyleGeneratedImage:: const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::StyleGradientImage::image const):

Canonical link: <a href="https://commits.webkit.org/262469@main">https://commits.webkit.org/262469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d854e6580eba9ff8a07a7d6b8031cdf0a38c334f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1543 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2460 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1687 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1524 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1637 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1463 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1556 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1390 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2297 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1376 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1313 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1406 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1416 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2468 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1295 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1383 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/399 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1504 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
```diff
@@ -138,29 +138,29 @@ void StyleGeneratedImage::computeIntrinsicDimensions(const RenderElement* render
     this->didRemoveClient(renderer);
 
-    if (m_clients.isEmpty())
+    if (!m_clients.computeSize())
```
isEmptyIgnoringNullReferences()

cdumez commented on Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
```diff
@@ -138,29 +138,29 @@ void StyleGeneratedImage::computeIntrinsicDimensions(const RenderElement* render
 void StyleGeneratedImage::addClient(RenderElement& renderer)
 {
-    if (m_clients.isEmpty())
+    if (!m_clients.computeSize())
```
isEmptyIgnoringNullReferences()

cdumez commented on Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
```diff
@@ -184,8 +184,10 @@ void StyleCrossfadeImage::imageChanged(CachedImage*, const IntRect*)
         return;
-    for (auto& client : clients().values())
-        client->imageChanged(this);
+    for (auto entry : clients()) {
```
Can we use `auto&` here or does it not build?

cdumez commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -162,8 +162,10 @@ void StyleFilterImage::imageChanged(CachedImage*, const IntRect*)
 
-    for (auto& client : clients().values())
-        client->imageChanged(static_cast<WrappedImagePtr>(this));
+    for (auto entry : clients()) {
```
Can we use `auto&` here or does it not build?

cdumez commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -162,8 +162,10 @@ void StyleFilterImage::imageChanged(CachedImage*, const IntRect*)
-        client->imageChanged(static_cast<WrappedImagePtr>(this));
+    for (auto entry : clients()) {
+        auto& client = entry.key;
+        client.imageChanged(static_cast<WrappedImagePtr>(this));
```
Personally, I would have kept on one line:
```
for (auto& client : clients())
    client.key.imageChanged(static_cast<WrappedImagePtr>(this));
```

Same above.

chirags27 replied:
yeah, that doesn't work:

```
/Volumes/Data/webkit-safari/OpenSource/Source/WebCore/rendering/style/StyleCanvasImage.cpp:120:16: error: loop variable 'entry' binds to a temporary value produced by a range of type 'const WeakHashCountedSet<WebCore::RenderElement>' [-Werror,-Wrange-loop-bind-reference]
    for (auto& entry : clients()) {
/Volumes/Data/webkit-safari/OpenSource/Source/WebCore/rendering/style/StyleCanvasImage.cpp:120:10: note: use non-reference type 'const WTF::WeakHashMap<WebCore::RenderElement, unsigned int>::PeekType'
```

cdumez reviewed (approved):
LGTM with comments.

chirags27 committed:
Move HashCountedSet to WeakHashCountedSet in StyleGeneratedImage
https://bugs.webkit.org/show_bug.cgi?id=254835
rdar://107480319

Reviewed by Chris Dumez.

Generated images should use a Weak container to keep track of
RenderElements so that we don't trigger UAF issues.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore:: const):
* Source/WebCore/rendering/style/StyleCanvasImage.cpp:
(WebCore::StyleCanvasImage::image const):
(WebCore::StyleCanvasImage::canvasChanged):
(WebCore::StyleCanvasImage::canvasResized):
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
(WebCore::StyleCrossfadeImage::imageChanged):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::imageChanged):
* Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
(WebCore::StyleGeneratedImage::addClient):
(WebCore::StyleGeneratedImage::removeClient):
(WebCore::StyleGeneratedImage::hasClient const):
* Source/WebCore/rendering/style/StyleGeneratedImage.h:
(WebCore::StyleGeneratedImage::clients const):
(WebCore::StyleGeneratedImage:: const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::StyleGradientImage::image const):

Canonical link: https://commits.webkit.org/262469@main
